File: telegram-bot-core/src/main/java/ru/tggc/telegrambotcore/registry/resolver/HandleParamResolver.java
package ru.tggc.telegrambotcore.registry.resolver;

import org.springframework.stereotype.Component;
import ru.tggc.telegrambotcore.annotation.params.HandleParam;
import ru.tggc.telegrambotcore.util.ParamConverter;

import java.lang.reflect.Parameter;
import java.util.regex.Matcher;

@Component
public class HandleParamResolver implements ParameterResolver<Object> {

    @Override
    public boolean supports(Parameter p) {
        return p.isAnnotationPresent(HandleParam.class);
    }

    @Override
    public Object resolve(Parameter parameter, HandlerCtx ctx) {
        Matcher matcher = ctx.matcher();

        if (matcher == null || !matcher.matches()) {
            return null;
        }

        String name = parameter.getAnnotation(HandleParam.class).value();

        String raw = matcher.group(name);

        return ParamConverter.convert(
                raw,
                parameter.getType()
        );
    }
}

File: telegram-bot-core/src/main/kotlin/ru/tggc/telegrambotcore/registry/resolver/HandleParamResolver.kt
package ru.tggc.telegrambotcore.registry.resolver

import org.springframework.stereotype.Component
import ru.tggc.telegrambotcore.annotation.params.HandleParam
import ru.tggc.telegrambotcore.util.ParamConverter
import java.lang.reflect.Parameter

@Component
class HandleParamResolver : ParameterResolver<Any?> {
    override fun supports(p: Parameter?): Boolean? = p?.isAnnotationPresent(HandleParam::class.java)

    override fun resolve(parameter: Parameter?, ctx: HandlerCtx): Any? =
        ctx.matcher
            ?.takeIf { it.matches() }
            ?.let { matcher ->
                val name = parameter
                    ?.getAnnotation(HandleParam::class.java)
                    ?.value
                    ?: return null

                ParamConverter.convert(
                    matcher.group(name),
                    parameter.type
                )
            }
}
